Expose CFunction

dpinte opened the pull request:
@flongford here is the first steps towards being able to use a CFunction. It's very close, I just need to get a full test running and add error management. Can you do a first review? The naming of the classes and modules definitely needs a second pass.

dpinte committed:
TST: split tests and cover some more cases

dpinte committed:
FEAT: expose the CFunction pointers to Cython

dpinte committed:
TST: add an example module to start experimenting with Cunction

dpinte committed:
REF: another pass on the Expression ctx and child classes

dpinte committed:
TST: forgot to add the BaseAcadoTestCase

dpinte committed:
FEAT: expose operator() and operator= for Expression

dpinte committed:
FIX: correct init for IntermediateState

dpinte committed:
FEAT: add Expression.__setitem___

dpinte committed:
FEAT: example module with a CFunction implemented on the C side

dpinte committed:
FEAT: export setUserData and operator() on CFunction

dpinte committed:
FEAT: new CFunction module

dpinte committed:
TST: basic test for the CFunction

dpinte committed:
FIX: function signature fix for expression_from_ref

dpinte committed:
REF: changed api extension to bindings extensino and add a cfunction extension

dpinte committed:
FEAT: add the tests supackage to the content of the package

flongford commented on setup.py:
```diff
@@ -1,9 +1,10 @@
 # Requires MACOSX_DEPLOYMENT_TARGET=10.9
 from setuptools import setup, Extension
 from Cython.Build import cythonize
+import numpy
```
`numpy` not currently included in requirements.txt

flongford commented on acadopy/tests/utils.py:
```diff
@@ -0,0 +1,15 @@
+import faulthandler
+import unittest
+
+from acadopy.api import clear_static_counters
```
Need to update to `acadopy.bindings`

dpinte replied:
done

flongford commented on acadopy/acado.pxd:
```diff
@@ -89,19 +94,27 @@ cdef extern from 'acado/symbolic_expression/expression.hpp' namespace 'ACADO':
+        IntermediateState(const DVector&)
 
     cdef cppclass TIME(ExpressionType):
         TIME()
```
Should `IntermediateState` be defined twice? (lines 112 and 125)

dpinte replied:
fixed

flongford commented on acadopy/bindings.pyx:
```diff
@@ -387,7 +394,25 @@ cdef class Expression:
+
+
 cdef class ExpressionType(Expression):
+    # Is it ok to let users rely on the Expression.__cinit__.
```
I suppose instantiating objects as `Expression` base classes makes more sense than abstracting the base class from an existing class...

flongford commented on acadopy/c_function.pyx:
```diff
@@ -0,0 +1,71 @@
+        
+        _result = new acado.Expression(_func(_exp))
+
+        # NOTE: do we need to make sure result keeps a reference on the PyFunction object 
```
Note: attempting to print an expression generated by this call yields:

`[ (real_t)(0)][ ][ C functions can not be printed , C functions can not be printed]`

in addition, attempting to minimise the expression in a 'OptimizationAlgorithm' raises:

`
AttributeError: 'acadopy.c_function.PyFunction' object has no attribute 'dim'
`

Which may suggest there's a benefit to removing the reference?

dpinte replied:
I've fixed a number of issues in the implementation. It's running now. The `dim` issue has been fixed

flongford reviewed (commented):
I've had an overview and tested some functionality of the `PyFunction` class, but will take a closer look next week.

dpinte committed:
REF: move acadopy.c_function to acadopy.function and removed examples

dpinte committed:
REF: update setup.py to reflect the module name changes and removal

dpinte committed:
FEAT: expose EvaluationPoint, add Function.evaluate

dpinte committed:
FEAT: add repr to DVector and copy constructor support

dpinte committed:
REF: update pxd file to reflect move of Function and DifferentialEquation

dpinte committed:
TST: update tests to reflec all changes

dpinte committed:
REF: remove the unicode management of c_string

dpinte committed:
TST: add missing import (+ some print statements to track what is going on)

dpinte committed:
CLN: cleanup

dpinte commented:
@flongford this should be functional now.

dpinte committed:
DEV: add numpy as a requirement

dpinte committed:
TST: use absolute imports

dpinte commented:
Tests segfault on Travis. I'll investigate

```Fatal Python error: Segmentation fault
Current thread 0x00007efde0e60740 (most recent call first):
  File "/src/acadopy/tests/test_function.py", line 121 in test_full_example
```

dpinte committed:
FIX: remove duplicated IntermediateState definition

dpinte commented:
> Tests segfault on Travis. I'll investigate 

The bug seems related to copying the DVector in the `setX` call. Will need further investigations.

flongford commented on acadopy/bindings.pyx:
```diff
@@ -404,24 +436,68 @@ cdef class DifferentialState(ExpressionType):
     """
 
-    def __cinit__(self, initialize=True):
+    def __cinit__(self, _arg=None, int nrows=1, int ncols=1, initialize=True):
```
Referencing elements of an `IntermediateState` other than for assignment raises:

`TypeError: 'acadopy.bindings.IntermediateState' object is not subscriptable`

This could be an issue if we want to use it as a container for generic acado-type objects to pass into a solver, since we may want to implement constraints:

  `a = IntermediateState(1)`
  `a[0] = Parameter()`
 `a_solver.subjectTo(lower_value <= a[0])`

But currently the `<=` operator is not supported for `IntermediateState` class.

dpinte replied:
__getitem__ is now implemented

dpinte committed:
FEAT: Expression __getitem__ implemented and fixed __setitem__

dpinte committed:
REF: improvements to PyFunction

- __cinit__ arguments match with C++ constructor
- expose a dim property
- remove print statements

dpinte committed:
TST: add tests for __getitem__

- and execute a full set/get item cycle

dpinte committed:
TST: improved test of PyFunction

- updated arguments in initializer
- add some dimension checks

dpinte committed:
DEV: add cpp files to .gitignore

dpinte commented:
One idea  about the segfault. The Travis testing  environment  uses acado 1.2.2beta while I am using the latest master on Mac.

dpinte commented:
Issue identified. It's related to the way `EvaluationPoint` extracts information from the underlying `Function`. It seems the `nx` of the Function is always 0, which leads to unassigned arrays and thus the segfault when copying the data (calling `set_x`). I'm investigating.
